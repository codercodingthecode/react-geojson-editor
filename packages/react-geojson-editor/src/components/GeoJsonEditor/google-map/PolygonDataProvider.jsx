/* eslint-disable react/no-did-update-set-state */
/* eslint-disable react/no-unused-state */

import React, { Component } from 'react';
import PropTypes from 'prop-types';

import { subtractPolygons } from '../../../turfjs-operations/subtraction';
import {
  convertPolygonToGeoJsonMulti, convertGeojsonToPolygon,
  convertGeojsonToGMData, gmDataToPolygonList,
  createGeoJson,
} from '../../../gm-operations/convert-polygon';
import { GeoJsonType } from '../../../custom-types';
import { GoogleMapContext } from './GoogleMapInitialzer';

/**
 * In charge of keeping track of polygons
 * @NOTE PolygonDataProvider is not allowed to call polygon methods (
 *      ex: polygon.setEditable or polygon.setMap)
 *      these calls should done outside
 */
export const PolygonDataContext = React.createContext({
  gmapData: null, // of type google.maps.Data, used to present existing polygons
  newPolygonList: [], // of type google.maps.Polygon
  polygonList: [], // converted from gmapData for editing purposes
  selectedPolygonList: [], // selected from either newPolygonList or polygonList

  // Actions
  // createGmapData: null, // of type Function
  addNewPolygon: null,
  // removeNewPolygon: null,
  selectPolygon: null,
  unSelectPolygon: null,

  // gmapData to polygonList
  // generatePolygonList: null,

  // Editor actions (applied on selected polygon) ??
  addAction: null, // combine selected polygons
  subtractAction: null,
  deleteAction: null,
  centerAction: null,

  discardChanges: null,
  persistChanges: null,
});
export class PolygonDataProvider extends Component {
      static contextType = GoogleMapContext;
      constructor(props, context) {
        super(props, context);
        // this.createGmapData = this.createGmapData.bind(this);
        this.addNewPolygon = this.addNewPolygon.bind(this);
        this.selectPolygon = this.selectPolygon.bind(this);
        this.unSelectPolygon = this.unSelectPolygon.bind(this);
        this.subtractAction = this.subtractAction.bind(this);
        this.deleteAction = this.deleteAction.bind(this);

        this.discardChanges = this.discardChanges.bind(this);
        this.persistChanges = this.persistChanges.bind(this);
        // this.generatePolygonList = this.generatePolygonList.bind(this);

        const gmapData = props.initialGeojson ?
          convertGeojsonToGMData(props.initialGeojson, context) :
          null;
        this.state = {
          gmapData,
          newPolygonList: [],
          polygonList: gmapData ? gmDataToPolygonList(gmapData, context) : [],
          selectedPolygonList: [],

          // Actions
          // createGmapData: this.createGmapData,
          addNewPolygon: this.addNewPolygon,
          selectPolygon: this.selectPolygon,
          unSelectPolygon: this.unSelectPolygon,
          // generatePolygonList: this.generatePolygonList,

          // Editor actions
          addAction: null,
          subtractAction: this.subtractAction,
          deleteAction: this.deleteAction,
          centerAction: null,

          discardChanges: this.discardChanges,
          persistChanges: this.persistChanges,
        };
      }

      // @TODO make sense to use getDerivedStateFromProps
      componentDidUpdate(prevProps) {
        const { initialGeojson } = this.props;

        if (prevProps.initialGeojson !== initialGeojson) {
          const gmapData = initialGeojson ?
            convertGeojsonToGMData(initialGeojson, this.context) :
            null;
          this.setState({
            gmapData,
            newPolygonList: [],
            polygonList: gmapData ? gmDataToPolygonList(gmapData, this.context) : [],
            selectedPolygonList: [],
          });
        }
      }

      // START Action handlers
      // createGmapData_NOT_IN_USE() {

      // }
      // generatePolygonList_NOT_IN_USE() {
      //   this.setState((state) => {
      //     const polygonList = gmDataToPolygonList(state.gmapData, this.context);
      //     return { polygonList };
      //   });
      // }

      addNewPolygon(polygon) {
        this.setState(state => ({
          newPolygonList: [...state.newPolygonList, polygon],
          selectedPolygonList: [...state.selectedPolygonList, polygon],
        }));
      }

      selectPolygon(polygon) {
        this.setState(state => ({
          selectedPolygonList: [...state.selectedPolygonList, polygon],
        }));
      }
      unSelectPolygon(polygon) {
        this.setState(state => ({
          selectedPolygonList: state.selectedPolygonList.filter(p => p !== polygon),
        }));
      }

      subtractAction(/* ev */) {
        // @TODO convertPolygonToGeoJson
        // @TODO convertPolygonToGeoJsonMulti
        // @TODO convertGeojsonToPolygon
        // @TODO convertGeojsonToPolygonMulti

        convertPolygonToGeoJsonMulti(this.state.selectedPolygonList, this.context)
          .then(geojsonList => subtractPolygons(geojsonList))
          .then(subtractedGeojson => convertGeojsonToPolygon(subtractedGeojson, this.context))
          .then((subtractedPolygon) => {
            this.setState(state => ({
              newPolygonList: [
                ...state.newPolygonList.filter(p => !state.selectedPolygonList.includes(p)),
                subtractedPolygon,
              ],
              polygonList: state.polygonList.filter(p => !state.selectedPolygonList.includes(p)),
              selectedPolygonList: [],
            }));
          });
      }

      deleteAction(/* ev */) {
        this.setState(state => ({
          newPolygonList: state.newPolygonList.filter(p =>
            !state.selectedPolygonList.includes(p)),
          polygonList: state.polygonList.filter(p =>
            !state.selectedPolygonList.includes(p)),
          selectedPolygonList: [],
        }));
      }

      discardChanges(_, cb) {
        this.setState(state => ({
          newPolygonList: [],
          polygonList: state.gmapData ? gmDataToPolygonList(state.gmapData, this.context) : [],
          selectedPolygonList: [],
        }), cb);
      }

      persistChanges(_, cb) {
        // 1. create a new GeoJson<MultiPolygon>
        // 2. send onSave action to the parent, telling it to
        //    update backend with newly created geojson
        createGeoJson([...this.state.newPolygonList, ...this.state.polygonList], this.context)
          .then(({ geojson, totalArea }) => {
            // console.log('[createGeoJson]', geojson, totalArea);

            cb(); // @TODO How do i save new GeoJson and turn off edit mode?
            // @TODO How do i save new GeoJson and turn off edit mode?
            this.props.onSave(geojson, totalArea);
          });
      }
      // END Action handlers

      render() {
        return (
          <PolygonDataContext.Provider value={this.state}>
            {this.props.children}
          </PolygonDataContext.Provider>
        );
      }
}


export default PolygonDataProvider;

PolygonDataProvider.propTypes = {
  children: PropTypes.node.isRequired,
  initialGeojson: GeoJsonType,
  onSave: PropTypes.func,
};

PolygonDataProvider.defaultProps = {
  initialGeojson: null,
  onSave: () => {},
};
