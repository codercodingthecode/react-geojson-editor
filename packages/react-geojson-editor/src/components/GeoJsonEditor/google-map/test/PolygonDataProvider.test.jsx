import React from 'react';

import { shallow, mount } from 'enzyme';
import { PolygonDataProvider } from '../PolygonDataProvider';

describe('Button', () => {
  it('has not changed', () => {
    const wrapper = shallow(<PolygonDataProvider><div /></PolygonDataProvider>);
    expect(wrapper).toMatchSnapshot();
  });
  it('renders by initializing methods to manage polygon storage', () => {
    const wrapper = mount(<PolygonDataProvider><div /></PolygonDataProvider>);
    expect(wrapper.state().addNewPolygon).toBeInstanceOf(Function);
    expect(wrapper.state().selectPolygon).toBeInstanceOf(Function);
    expect(wrapper.state().unSelectPolygon).toBeInstanceOf(Function);
  });
  // @TODO explain why addNewPolygon function is added to the state object
});
