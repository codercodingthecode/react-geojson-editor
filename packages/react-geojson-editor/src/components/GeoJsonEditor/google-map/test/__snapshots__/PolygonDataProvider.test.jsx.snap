// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Button has not changed 1`] = `
ShallowWrapper {
  Symbol(enzyme.__root__): [Circular],
  Symbol(enzyme.__unrendered__): <PolygonDataProvider
    initialGeojson={null}
    onSave={[Function]}
  >
    <div />
  </PolygonDataProvider>,
  Symbol(enzyme.__renderer__): Object {
    "batchedUpdates": [Function],
    "checkPropTypes": [Function],
    "getNode": [Function],
    "render": [Function],
    "simulateError": [Function],
    "simulateEvent": [Function],
    "unmount": [Function],
  },
  Symbol(enzyme.__node__): Object {
    "instance": null,
    "key": undefined,
    "nodeType": "function",
    "props": Object {
      "children": <div />,
      "value": Object {
        "addAction": null,
        "addNewPolygon": [Function],
        "centerAction": null,
        "deleteAction": [Function],
        "discardChanges": [Function],
        "gmapData": null,
        "newPolygonList": Array [],
        "persistChanges": [Function],
        "polygonList": Array [],
        "selectPolygon": [Function],
        "selectedPolygonList": Array [],
        "subtractAction": [Function],
        "unSelectPolygon": [Function],
      },
    },
    "ref": null,
    "rendered": Object {
      "instance": null,
      "key": undefined,
      "nodeType": "host",
      "props": Object {},
      "ref": null,
      "rendered": null,
      "type": "div",
    },
    "type": Object {
      "$$typeof": Symbol(react.provider),
      "_context": Object {
        "$$typeof": Symbol(react.context),
        "Consumer": Object {
          "$$typeof": Symbol(react.context),
          "_calculateChangedBits": null,
          "_context": [Circular],
        },
        "Provider": [Circular],
        "_calculateChangedBits": null,
        "_currentRenderer": null,
        "_currentRenderer2": null,
        "_currentValue": Object {
          "addAction": null,
          "addNewPolygon": null,
          "centerAction": null,
          "deleteAction": null,
          "discardChanges": null,
          "gmapData": null,
          "newPolygonList": Array [],
          "persistChanges": null,
          "polygonList": Array [],
          "selectPolygon": null,
          "selectedPolygonList": Array [],
          "subtractAction": null,
          "unSelectPolygon": null,
        },
        "_currentValue2": Object {
          "addAction": null,
          "addNewPolygon": null,
          "centerAction": null,
          "deleteAction": null,
          "discardChanges": null,
          "gmapData": null,
          "newPolygonList": Array [],
          "persistChanges": null,
          "polygonList": Array [],
          "selectPolygon": null,
          "selectedPolygonList": Array [],
          "subtractAction": null,
          "unSelectPolygon": null,
        },
        "_threadCount": 0,
      },
    },
  },
  Symbol(enzyme.__nodes__): Array [
    Object {
      "instance": null,
      "key": undefined,
      "nodeType": "function",
      "props": Object {
        "children": <div />,
        "value": Object {
          "addAction": null,
          "addNewPolygon": [Function],
          "centerAction": null,
          "deleteAction": [Function],
          "discardChanges": [Function],
          "gmapData": null,
          "newPolygonList": Array [],
          "persistChanges": [Function],
          "polygonList": Array [],
          "selectPolygon": [Function],
          "selectedPolygonList": Array [],
          "subtractAction": [Function],
          "unSelectPolygon": [Function],
        },
      },
      "ref": null,
      "rendered": Object {
        "instance": null,
        "key": undefined,
        "nodeType": "host",
        "props": Object {},
        "ref": null,
        "rendered": null,
        "type": "div",
      },
      "type": Object {
        "$$typeof": Symbol(react.provider),
        "_context": Object {
          "$$typeof": Symbol(react.context),
          "Consumer": Object {
            "$$typeof": Symbol(react.context),
            "_calculateChangedBits": null,
            "_context": [Circular],
          },
          "Provider": [Circular],
          "_calculateChangedBits": null,
          "_currentRenderer": null,
          "_currentRenderer2": null,
          "_currentValue": Object {
            "addAction": null,
            "addNewPolygon": null,
            "centerAction": null,
            "deleteAction": null,
            "discardChanges": null,
            "gmapData": null,
            "newPolygonList": Array [],
            "persistChanges": null,
            "polygonList": Array [],
            "selectPolygon": null,
            "selectedPolygonList": Array [],
            "subtractAction": null,
            "unSelectPolygon": null,
          },
          "_currentValue2": Object {
            "addAction": null,
            "addNewPolygon": null,
            "centerAction": null,
            "deleteAction": null,
            "discardChanges": null,
            "gmapData": null,
            "newPolygonList": Array [],
            "persistChanges": null,
            "polygonList": Array [],
            "selectPolygon": null,
            "selectedPolygonList": Array [],
            "subtractAction": null,
            "unSelectPolygon": null,
          },
          "_threadCount": 0,
        },
      },
    },
  ],
  Symbol(enzyme.__options__): Object {
    "adapter": ReactSixteenAdapter {
      "options": Object {
        "enableComponentDidUpdateOnSetState": true,
        "legacyContextMode": "parent",
        "lifecycles": Object {
          "componentDidUpdate": Object {
            "onSetState": true,
          },
          "getChildContext": Object {
            "calledByRenderer": false,
          },
          "getDerivedStateFromProps": true,
          "getSnapshotBeforeUpdate": true,
          "setState": Object {
            "skipsComponentDidUpdateOnNullish": true,
          },
        },
      },
    },
    "disableLifecycleMethods": true,
  },
}
`;
